types/src/trusted_state.rs

```diff
@@ -0,0 +1,195 @@
+};
+use anyhow::{format_err, Result};
+use std::sync::Arc;
+
+/// `TrustedState` keeps track of our latest trusted state, including the latest
+/// verified version and the latest verified validator set.
+#[derive(Clone, Debug)]
+pub struct TrustedState {
+    /// The latest verified version of either a waypoint or a ledger info, either
+    /// inside an epoch or the epoch change ledger info. If the TrustedState is
+    /// generated from an initial waypoint, the latest_version is coincidentally
+    /// the same as the waypoint version.
+    latest_version: Version,
+    /// The current verifier. If we're starting up fresh, this is probably a
+    /// waypoint from our config. Otherwise, this is generated from the validator
+    /// set in the last known epoch change ledger info.
+    verifier: VerifierType,
+}
+
+/// `TrustedStateChange` is the result of attempting to ratchet to a new trusted
+/// state. In order to reduce redundant error checking, `TrustedStateChange` also
+/// contains references to relevant items used to ratchet us.
+#[derive(Clone, Debug)]
+pub enum TrustedStateChange<'a> {
+    /// We have a newer `TrustedState` but it's still in the same epoch, so only
+    /// the latest trusted version changed.
+    Version {
+        new_state: TrustedState,
+        latest_li: &'a LedgerInfoWithSignatures,
+    },
+    /// We have a newer `TrustedState` and there was at least one epoch change,
+    /// so we have a newer trusted version and a newer trusted validator set.
+    Epoch {
+        new_state: TrustedState,
+        latest_li: &'a LedgerInfoWithSignatures,
+        latest_epoch_change_li: &'a LedgerInfoWithSignatures,
+        latest_validator_set: &'a ValidatorSet,
+    },
+    /// Our current `TrustedState` is more recent.
+    Stale,
+}
+
+impl TrustedState {
+    /// Create an initial trusted state from a waypoint.
+    pub fn from_waypoint(waypoint: Waypoint) -> Self {
+        Self {
+            latest_version: waypoint.version(),
+            verifier: VerifierType::Waypoint(waypoint),
+        }
+    }
+
+    /// Create an initial trusted state that will trust the first genesis
+    /// presented to it.
+    ///
+    /// WARNING: this is obviously unsafe, as a malicious peer could present any
+    /// arbitrary genesis and this TrustedState would gladly accept it.
+    // TODO(philiphayes/dmitrip): remove this when waypoints are completely
+    // integrated with client code.
+    #[allow(non_snake_case)]
+    pub fn new_trust_any_genesis_WARNING_UNSAFE() -> Self {
+        Self {
+            latest_version: 0,
+            verifier: VerifierType::TrustedVerifier(EpochInfo::empty()),
+        }
+    }
+
+    /// Create an initial trusted state from an epoch change ledger info.
+    pub fn from_epoch_change_ledger_info(epoch_change_li: &LedgerInfo) -> Result<Self> {
+        // Generate the EpochInfo from the new validator set.
+        let epoch_info = EpochInfo {
+            epoch: epoch_change_li.epoch() + 1,
```

Comment:
@zekun000 as we discussed before, the version where the next_validator_set() shows up non-None, is regarded still in the last epoch, and that's the server logic. It's different here... (Thanks @phlip9 for pointing that out.)

Reply:
o .. is it less of a problem now because we no longer have empty blocks?

Reply:
The logic is about when the server returns the ValidatorChangeProof. It's the server determining whether the server.epoch > client.epoch and returns the proof.

If the server is at the edge of epoch change, it still considers itself in the last epoch and would not return the proof to client, also the LedgerInfo with Some(next_validator_set) is at last epoch and would not trigger the `epoch_change_verification_required`.

I think it's consistent between server and client, if the server returns an epoch change ledger info means the server believes it's in next epoch. Also the client will only move the epoch further when the server shows a ledger info in next epoch(epoch change li is not).
Hope this makes sense.

Reply:
If the server is at the edge of an epoch change, shouldn't it consider itself in that next epoch?

Reply:
good point, storage differentiate it for consensus with client